### PR TITLE
grpc: Add join Dial Option

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds a join Dial Option. It also contains tests for this join Dial Option and the already present join Server Option.

RELEASE NOTES: N/A

### CODE REVIEW

File: default_dial_option_server_option_test.go [hunk 1]
```diff
@@ -26,6 +26,9 @@ import (
 	"google.golang.org/grpc/internal"
 )
 
+const maxRecvSize = 998765
+const initialWindowSize = 100
```
Comment (dfawley) [inline default_dial_option_server_option_test.go:30]: Please make these local inside their tests.

File: internal/internal.go [hunk 2]
```diff
@@ -77,6 +77,9 @@ var (
 	// ClearGlobalDialOptions clears the array of extra DialOption. This
 	// method is useful in testing and benchmarking.
 	ClearGlobalDialOptions func()
+	// JoinDialOptions combines the dial options passed as arguments into a
+	// single dial option.
+	JoinDialOptions interface{}
```
Comment (dfawley) [inline internal/internal.go:82]: Please comment: `// func(...grpc.DialOption) grpc.DialOption`

**Review by dfawley — APPROVED** [review]
LGTM except the two nits.

File: internal/internal.go (hunk 2 above)
Comment (zasweq) [reply internal/internal.go:82]: Done.

File: default_dial_option_server_option_test.go (hunk 1 above)
Comment (zasweq) [reply default_dial_option_server_option_test.go:30]: This is used 4/2 times, respectively?

Comment (zasweq) [reply default_dial_option_server_option_test.go:30]: Actually this is in grpc I see I'll move to tests.

**Pull request merged**